File: Research/Cluster_py/ctrlclusterf.f90
!-------------------------
!
! Calcurate cluster when call from python script 
!
!-------------------------
module fortcluster
  use ISO_C_binding
  implicit none
  type atom
    character(5) :: atomname
    integer iz
    double precision x,y,z
  end type atom
contains

  double precision function veclength(vec)
    double precision,intent(in) :: vec(3)
    integer j
    veclength=0.
    do j=1,3
      veclength = veclength + vec(j)**2
    end do
    veclength=sqrt(veclength)
  end function veclength
  
  subroutine baburusort(allatom,natoms)
    implicit none
    type(atom),intent(inout) :: allatom(:)
    integer,intent(in) :: natoms
    logical :: baburu=.true.
    integer :: baburu_c=0
    integer i
    double precision karix ,kariy ,kariz
    integer kariiz
    character(5) ::  kariname

    do while(baburu)
      baburu_c=baburu_c+1
      baburu=.false.
      do i=1,natoms-baburu_c
        if (allatom(i)%x**2+allatom(i)%y**2+allatom(i)%z**2 > &
           & allatom(i+1)%x**2+allatom(i+1)%y**2+allatom(i+1)%z**2) then
           kariname=allatom(i+1)%atomname
           kariiz=allatom(i+1)%iz
           karix=allatom(i+1)%x
           kariy=allatom(i+1)%y
           kariz=allatom(i+1)%z
           allatom(i+1)%atomname=allatom(i)%atomname
           allatom(i+1)%iz=allatom(i)%iz
           allatom(i+1)%x=allatom(i)%x
           allatom(i+1)%y=allatom(i)%y
           allatom(i+1)%z=allatom(i)%z
           allatom(i)%atomname=kariname
           allatom(i)%iz=kariiz
           allatom(i)%x=karix
           allatom(i)%y=kariy
           allatom(i)%z=kariz
           baburu=.true.
        end if
      end do
    end do
  end subroutine baburusort         
end module fortcluster

subroutine cluster(clradius,Plat,origincart,nbas,cart,alat,alliz)
  use fortcluster
  implicit none
  double precision,parameter :: FACTOR=5
  double precision,parameter :: BOHR=0.529177
  double precision,intent(in) :: clradius
  double precision,intent(in) :: Plat(3,3)
  double precision,intent(in) :: origincart(3)
  integer,intent(in) :: nbas
  double precision,intent(in) :: cart(nbas,3)
  double precision,intent(in) :: alat
  integer,intent(in) :: alliz(nbas)
  double precision cutoff
  integer i0max , i1max , i2max
  integer i0 , i1 , i2 , k ,j
  integer :: natoms=0
  integer :: n=1
  double precision y(3)
  integer nbas2
  integer lmax
  type(atom),allocatable :: allclass(:)
  type(atom),allocatable :: allatom(:)
  
  
  open(17,file="cluster.in",status="old")
  do j=1,5
   read(17,*)
  end do

  read(17,*) nbas2

  if (nbas/=nbas2)then
    print *,"error1"
    stop 
  end if

 allocate(allclass(nbas))
 
  do j=1,nbas
    read (17,*) allclass(j)%atomname ,&
               &  allclass(j)%x , allclass(j)%y , allclass(j)%z
  end do
  
  close(17)

  print*,""
  print*,"allsite"
  do j=1,nbas
    write (*,40) allclass(j)%atomname,allclass(j)%x,allclass(j)%y,allclass(j)%z
  end do
  40 format(A4," ",F10.6," ",F10.6," "F10.6) 

  cutoff = clradius
  i0max = cutoff * FACTOR / veclength(Plat(1,:))
  i1max = cutoff * FACTOR / veclength(Plat(2,:))
  i2max = cutoff * FACTOR / veclength(Plat(3,:))
  
  write (*,50) "i0max=",i0max,"i1max=",i1max,"i2max=",i2max
  50 format(A6,I3," ",A6,I3," ",A6,I3)
  do i0=-i0max,i0max
    do i1=-i1max,i1max
      do i2=-i2max,i2max
        do k=1,nbas
          do j=1,3
            y(j)=i0*Plat(1,j)+i1*Plat(2,j)+i2*plat(3,j) &
                 &  +cart(k,j) - origincart(j)
            y(j) = y(j)* alat * BOHR 
          end do
          if (veclength(y)< clradius) then
            natoms=natoms+1
          end if
        end do
      end do
    end do
  end do

  write (*,51) "natoms=",natoms 
  51 format(A7,I4)

  allocate(allatom(natoms))

  do i0=-i0max,i0max
    do i1=-i1max,i1max
      do i2=-i2max,i2max
        do k=1,nbas
          do j=1,3
            y(j)=i0*Plat(1,j)+i1*Plat(2,j)+i2*plat(3,j) &
                  & +cart(k,j) - origincart(j)
            y(j) = y(j)* alat * BOHR 
          end do
          if (veclength(y)< clradius) then
            allatom(n)%atomname=allclass(k)%atomname
            allatom(n)%iz=alliz(k)
            allatom(n)%x=y(1)
            allatom(n)%y=y(2)
            allatom(n)%z=y(3)
            n=n+1
          end if
        end do
      end do
    end do
  end do
  
  print* ,"coordinate" 
  do j=1,natoms
    write (*,101) allatom(j)%atomname,allatom(j)%x,allatom(j)%y,allatom(j)%z 
  end do
  
  call baburusort(allatom,natoms)

  print*,"sort coordinate"
  do j=1,natoms
    write (*,101) allatom(j)%atomname,allatom(j)%x,allatom(j)%y,allatom(j)%z 
  end do
         
  open(18,file="cluster.xyz",status="replace")
  
  write (18,100) natoms
  write (18,*)
  do j=1,natoms
    write (18,101) allatom(j)%atomname,allatom(j)%x,allatom(j)%y,allatom(j)%z 
  end do
  100 format(I4)
  101 format(A4," ",F10.6," ",F10.6," "F10.6) 
  close(18)

  open(19,file="instr",status="replace")
  write(19,102) natoms , natoms , 2
  do j=1,natoms
    if (allatom(j)%iz<=18) then 
      lmax=1
    else 
      lmax=2
    end if 
    write(19,103) allatom(j)%atomname,lmax,allatom(j)%iz, &
                  & allatom(j)%x/BOHR,allatom(j)%y/BOHR,allatom(j)%z/BOHR
  
  end do  
  102 format(I5,I5,I5)
  103 format(A4," ",I1," ",I2,"  ",F10.6," ",F10.6," ",F10.6)

  deallocate(allclass)
  deallocate(allatom)
end subroutine cluster

File: Research/Cluster_py/ctrlclusterf.py
"""
To calculate cluster.xyz from CTRL(lmtofile)

call fortran sub routine

"""
import sys
import numpy as np
import copy
from cProfile import Profile
import time
from fortcall_clus import Fort

BOHR=0.529177
FACTOR=5

def main():
  with open ("CTRL",mode="r") as fp:
    site_t=None
    class_t=None
    plat_n=0
    for line in fp:
      word = line.split()
      if word[0]=="STRUC":  
        worde = line.split("=")
        alat=float(worde[1].strip())
        plat_n=4
      if word[0]=="DIM":
        for w in word:
          worde = w.split("=")
          if worde[0]=="NBAS":
            nbas=int(worde[1])
          if worde[0]=="NCLASS":
            nclass=int(worde[1])
#  print(alat,nbas)
      if plat_n==3:
        Plat=np.zeros((3,3))
        worde = line.split("=")
        worde2 = worde[1].split()
        Plat[0][0]=float(worde2[0])
        Plat[0][1]=float(worde2[1])
        Plat[0][2]=float(worde2[2])
      elif plat_n==2:
        Plat[1][0]=float(word[0])
        Plat[1][1]=float(word[1])
        Plat[1][2]=float(word[2])
      elif plat_n==1:
        Plat[2][0]=float(word[0])
        Plat[2][1]=float(word[1])
        Plat[2][2]=float(word[2])
      plat_n-=1 
      if word[0]=="CLASS":
        class_t=True
        class_n=nclass
        allclass_ctrl=[]
      if class_t:
        if len(word)<5:
          continue
        worde=line.split("=")
        class_ctrl=Class_ctrl(worde[1].split()[0],worde[2].split()[0],worde[3].split()[0])
        allclass_ctrl.append(class_ctrl)
        class_n-=1  
        if class_n==0:
          class_t=False  
      if word[0]=="SITE":
        site_t=True
        site_n=nbas 
        allatom=[]
      if site_t:
        atom=Atom()
        worde = line.split("=")
        atom.name=worde[1].split()[0]
        atom.pos=np.array([float(worde[2].split()[0]),float(worde[2].split()[1]),float(worde[2].split()[2])])
        allatom.append(atom)
        site_n-=1
        if site_n==0:
          site_t=False
      if site_t==False:
        break

  for i in allatom:
    i.ctrltuika(allclass_ctrl)  

  print ("\nallatom\n")

  for i in range (len(allatom)):
    print (allatom[i].name,"iz=",allatom[i].ze,"rad=",allatom[i].rad,"pos=",allatom[i].pos)

  print("")

  print("Are There unnecessary atoms? yes or no")
  atomcheck=input()
  kariallatom=[]

  while True:

    if atomcheck=="yes" or atomcheck=="y":
      print("Plese input unnecessary atoms")
      unatoms=input().split()
      for i in range(len(allatom)):
        unatom_c=0
        for unatom in unatoms:
          if allatom[i].name==unatom:
            unatom_c+=1
        if unatom_c==0:
          kariallatom.append(allatom[i])
      allatom=copy.deepcopy(kariallatom)
      kariallatom=[]
      for i in range (len(allatom)):
        print (allatom[i].name,"iz=",allatom[i].ze,"rad=",allatom[i].rad,"pos=",allatom[i].pos)
      print("")
      print("It\'s OK? yes or no")
      atomok=input()
      if atomok=="yes" or atomok=="y": 
        break
  
 
    elif atomcheck=="no" or atomcheck=="n":
      break      

    else:
      print("Plese input yes or no")
      atomcheck=input()

  nbas=len(allatom)
#  print(Plat)
#  sys.exit()
#  Pos_np=np.array(Pos)
#  print(Atom)
#  print(Pos_np)   
#  Plat=Plat*alat*BOHR
#  Pos_np=Pos_np*alat*BOHR
#  print(Pos_np)   
  
  print("where is center ?(Atomname or Coordinate)")
  centeratom=input().split()
  c=-1
  origincart=np.array([0.,0.,0.])
  if len (centeratom) ==1:
    for i in range(nbas):
      if centeratom[0]==allatom[i].name:
        c=i
        break
    origincart=allatom[c].pos
    if c==-1:
      print("error : There is no atom of that name")
      sys.exit()
  elif len(centeratom)==3:
    for i in range(3):
      origincart[i]=float(centeratom[i])
  else:
    print("error")
    sys.exit()
    
  
   
  print("How is clustersize ? (please input cluster radius)")
  clradius=float(input()) 
  
  with open ("cluster.in",mode="w") as clin:
    for i in range(3):
      clin.write(str("{:>11}".format("{:.8f}".format(Plat[i][0])))) 
      clin.write(str("{:>11}".format("{:.8f}".format(Plat[i][1])))) 
      clin.write(str("{:>11}".format("{:.8f}".format(Plat[i][2])))) 
      clin.write("\n")

    clin.write(" 0 0 0\n")
    clin.write(str("{:>12}".format("{:.8f}".format(allatom[c].pos[0]))))
    clin.write(str("{:>12}".format("{:.8f}".format(allatom[c].pos[1]))))
    clin.write(str("{:>12}".format("{:.8f}".format(allatom[c].pos[2]))))
    clin.write("\n")
    clin.write(" "+str(nbas)+"\n")
    for i in range(nbas):   
      clin.write(str("{:>3}".format(allatom[i].name)))
      clin.write(str("{:>12}".format("{:.8f}".format(allatom[i].pos[0]))))
      clin.write(str("{:>12}".format("{:.8f}".format(allatom[i].pos[1]))))
      clin.write(str("{:>12}".format("{:.8f}".format(allatom[i].pos[2]))))
      clin.write("\n")
    clin.write(str("{:.4f}".format(alat*BOHR))+"\n")
    for i in range(3):
      clin.write(str(clradius)+"\n")

#  sys.exit()
  cart=np.zeros((nbas,3))
  alliz=np.zeros((nbas),dtype="int32")
  for i in range(nbas):
    alliz[i]=allatom[i].ze
    for j in range(3):
      cart[i][j]=allatom[i].pos[j]
  
  fort=Fort()
  fort.call_fortran(clradius,Plat,origincart,nbas,cart,alat,alliz)

  print("Succeeded!")
 
  """
  cutoff=clradius
  i0max=int(cutoff * FACTOR /np.sqrt(np.sum(Plat[0]**2)))
  i1max=int(cutoff * FACTOR /np.sqrt(np.sum(Plat[1]**2)))
  i2max=int(cutoff * FACTOR /np.sqrt(np.sum(Plat[2]**2)))

  clusteratom=[]
  y=np.array([0.,0.,0.])
  n=0  
  pr=Profile()
  pr.enable()
  t1=time.time()
  for i0 in range(-i0max,i0max+1):
    for i1 in range(-i1max,i1max+1):
      for i2 in range(-i2max,i2max+1):
        for k in range(nbas):
          y=i0*Plat[0] + i1*Plat[1] + i2*Plat[2] + allatom[k].pos - origincart
          y*=alat*BOHR
          if np.sqrt(sum(y**2))<clradius:
            atom=Atom()
            atom.name=allatom[k].name
            atom.ze=allatom[k].ze
            atom.pos=copy.deepcopy(y)
            clusteratom.append(atom)
            n+=1
  print(n)
  t2=time.time()
  print(t2-t1)
  sys.exit()
  pr.disable()
  pr.print_stats()
  sys.exit()
  for i in range(len(clusteratom)):
    print(clusteratom[i].name,clusteratom[i].ze,clusteratom[i].pos) 
  """
class Class_ctrl():
  def __init__(self,name,ze,rad):
    self.name=name
    self.ze=int(ze)
    self.rad=float(rad)


class Atom():
  def __init__(self):
    self.name=None
    self.pos=None
    self.ze=None
    self.rad=None

  def ctrltuika(self,allclass_ctrl):
    for i in allclass_ctrl:
      if i.name==self.name:
        self.ze=i.ze
        self.rad=i.rad
        break



if __name__=="__main__":
  main()
